ARCHITECTURE.md

```diff
@@ -40,27 +40,174 @@ The Python code in HOOMD-blue is mostly written to wrap core functionality
+
+In Python, many base classes exist to facilitate code reuse. This leads to very
+lean user facing classes, but requires more fundamental understanding of the model
+to address cases where customization is necessary, or changes are required. The
+first aspect of the data model discussed is that of most HOOMD operations or
+related classes.
+
+- `Operation`
+    - `_HOOMDBaseObject`
+        - `_HOOMDGetSetAttrBase`
+        - `_DependencyRelation`
+
+#### `_DependencyRelation`
+
+`_DependencyRelation` helps define a dependent dependency relationship between
+two objects. The class inherited by `_HOOMDBaseObject` to handle dependent
+relationships between operations in Python. This feature has yet to be used to
+much extent.  See `hoomd/operation.py` for the source code.
+
+#### `_HOOMDGetSetAttrBase`
+
+`_HOOMDGetSetAttrBase` provides hooks for exposing object attributes through two
+internal (underscored) attributes `_param_dict` and `_typeparam_dict`.
+`_param_dict` is an instance of type `hoomd.data.parameterdicts.ParameterDict`,
+and `_typeparam_dict` is a dictionary of attribute names to
+`hoomd.data.typeparam.TypeParameter`. This serves as the fundamental way of
+specifying object attributes in Python. The class provides a multitude of hooks
+to enable custom attribute querying and setting when necessary. See
+`hoomd/operation.py` for the source code and the sections on `TypeParameter`'s
+and `ParameterDict`'s for more information.
+
+#### `_HOOMDBaseObject`
+
+`_HOOMDBaseObject` combines `_HOOMDGetSetAttrBase` and `_DependencyRelation` to
+provide dependency handling, validated and processed attribute setting,
+pickling, and pybind11 C++ class syncing in an automated and structured way.
+Most methods unique to or customized from base classes revolve around allowing
+`_param_dict` and `_typeparam_dict` to sync to and from C++ when attached and
+not when unattached. See `hoomd/operation.py` for source code.
+
+### Attribute Validation and Defaults
+
+In Python, five fundamental classes exist for value validation, processing, and
+syncing when attached: `hoomd.data.parameterdicts.ParameterDict`,
+`hoomd.data.parameterdicts.TypeParameterDict`,
+`hoomd.data.parameterdicts.AttachedTypeParameterDict`,
+`hoomd.data.typeparam.TypeParameter`, and
+`hoomd.data.syncedlist.SyncedList`.
+These can be/are used by `_HOOMDBaseObject` subclasses as well as others.
+
+#### `ParameterDict`
+
+`ParameterDict` provide a mapping (`dict`) interface from attribute names to
+validated and processed values. Each instance of `ParameterDict` has its own
+specification defining the validation logic for its keys. `_HOOMDBaseObject`
+subclasses automatically sync the `ParameterDict` instance in the `_param_dict`
+attribute.
+
+This class should be used to define all attributes shared with C++ member
+variables that are on a per-object basis (i.e. not per type). A good example of
+this class can be seen in many HOOMD classes, but a good example is
```

Comment:
```suggestion
variables that are on a per-object basis (i.e. not per type). Examples of
this can be seen in many HOOMD classes.  One good example is
```